### PR TITLE
[Merged by Bors] - feat: `positivity` extension for `ite`

### PR DESCRIPTION
**Pull request by YaelDillies**
Match https://github.com/leanprover-community/mathlib/pull/17650

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 1]
```diff
@@ -19,6 +19,73 @@ This file sets up the basic `positivity` extensions tagged with the `@[positivit
...
+  ite p a b ≠ 0 :=
+ite_ne_zero _ ha hb.ne'
+
+end ite
+
+/-- The `positivity` extension which identifies expressions of the form `ite p a b`,
+such that `positivity` successfully recognises both `a` and `b`. -/
+@[positivity ite _ _] def evalIte : PositivityExt where eval {u α} zα pα e := do
+  let .app (.app (.app f (p : Q(Prop))) (a : Q($α))) (b : Q($α)) ← withReducible (whnf e)
+    | throwError "not ite"
```
Comment (YaelDillies) [inline Mathlib/Tactic/Positivity/Basic.lean:53]: I think I messed up this pattern-matching, but I can't figure out how to fix it. @digama0, would you enlighten me?

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 2]
```diff
@@ -19,6 +19,73 @@ This file sets up the basic `positivity` extensions tagged with the `@[positivit
...
+  let ra ← core zα pα a; let rb ← core zα pα b
+  let _a ← synthInstanceQ (q(Preorder $α) : Q(Type u))
```
Comment (YaelDillies) [inline Mathlib/Tactic/Positivity/Basic.lean:55]: Synthesizing `Preorder α` that early means the extension is less general than its Lean 3 counterpart. Should I/can I move this line into each branch of the big `ra, rb` pattern-match, with an appropriate tweak to this instance I ask synthesis of?

**Review by YaelDillies — COMMENTED** [review]
Help wanted!

File: Mathlib/Tactic/Positivity/Basic.lean (hunk 1 above)
Comment (dwrensha) [reply Mathlib/Tactic/Positivity/Basic.lean]: This needed an additional `.app`, to pick up the `Decidable` instance.

File: Mathlib/Tactic/Positivity/Basic.lean (hunk 2 above)
Comment (dwrensha) [reply Mathlib/Tactic/Positivity/Basic.lean]: I went ahead and pushed the `synthInstanceQ` call into the branches that needed it.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**